File: 250324/메두사와 전사들/medusa-and-warriors.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        sy = Integer.parseInt(st.nextToken());
        sx = Integer.parseInt(st.nextToken());
        ey = Integer.parseInt(st.nextToken());
        ex = Integer.parseInt(st.nextToken());
        
        medusa = new Medusa(sy,sx);
        
        st = new StringTokenizer(br.readLine());
        warriors = new Warrior[M];
        wBoard = new int[N][N];
        for(int i = 0; i < M; i++) {
        	int y = Integer.parseInt(st.nextToken());
        	int x = Integer.parseInt(st.nextToken());
        	warriors[i] = new Warrior(i, y,x);
        	wBoard[y][x]++;
        }
        
        board = new int[N][N];
        dist = new int[N][N];
        for(int y=0; y<N; y++) {
        	st = new StringTokenizer(br.readLine());
        	for(int x = 0; x<N; x++) {
        		board[y][x] = Integer.parseInt(st.nextToken());
        		dist[y][x] = INF;
        	}
        }
        
        init();
        
        if(dist[sy][sx] == INF) {
        	System.out.println(-1);
        	return;
        }
        
//        medusa.move();
        
        
        StringBuilder sb = new StringBuilder();
        while(true) {
        	medusa.move();
        	if(medusa.end()) {
        		sb.append(0).append('\n');
        		break;
        	}
        	Sight bestSight = medusa.findBestSight();
        	
        	int[] moved = moveAll(bestSight);
        	sb.append(moved[0]+" "+bestSight.cnt+" "+moved[1]).append('\n');
        }
        System.out.println(sb);
    }
    
    static int[] moveAll(Sight bestSight){
    	int sighted[][] = bestSight.sighted;
    	List<Warrior> notFreezed = bestSight.notFreezed;
    	
    	int distSum = 0;
    	int attackerCnt = 0;
    	
    	for(Warrior w : notFreezed) {
    		Move move = w.move(sighted);
    		if(move.attack)
    			attackerCnt++;
    		distSum += move.dist;
    	}
    	
    	
    	
    	return new int[] {distSum, attackerCnt};
    }
    
    
    static int N,M;
    static StringTokenizer st;
    static Medusa medusa;
    static int dist[][];
    static Warrior warriors[];
    static int sy,sx,ey,ex;
    
    static class Pair{
    	int y,x;
    	public Pair(int y,int x) {
    		this.y = y;
    		this.x = x;
    	}
    	
    	public String toString() {
    		return String.format("(%d, %d)", y,x);
    	}
    	
    }
    static Pair NO_PAIR = new Pair(-1,-1);
    
    
    static class Medusa{
    	int y,x;
    	public Medusa(int y,int x) {
    		this.y = y;
    		this.x = x;
    	}
    	
    	public boolean isSame(int y,int x) {
    		return this.y == y && this.x == x;
    	}
    	
    	public boolean end() {
    		return isSame(ey,ex);
    	}
    	
    	public String toString() {
    		return String.format("(%d, %d)", y,x);
    	}
    	
    	/*
    	 * 메두사의 이동 
    	 */
    	public void move() {
    		Pair nxt = findNxtPos();
    		y = nxt.y;
    		x = nxt.x;
    		
    		for(int i = 0; i < M;i++) {
    			if(warriors[i] == NO_WARRIOR)
    				continue;
    			int wy = warriors[i].y;
    			int wx = warriors[i].x;
    			if(isSame(wy,wx)) { // 전사가 있으면 공격받고 사라짐. 
    				wBoard[wy][wx]--;
    				warriors[i] = NO_WARRIOR;
    			}
    		}
    	}
    	
    	/*
    	 * 다음 위치 정하기 
    	 */
    	private Pair findNxtPos() {
    		int curDist = dist[y][x];
    		Pair ret = NO_PAIR;
    		for(int dir = 0; dir < 4; dir++) { // 상하좌우 우선순위 
    			int ny = y + dy[dir];
    			int nx = x + dx[dir];
    			
    			if(OOB(ny,nx) || board[ny][nx] == 1)
    				continue;
    			
    			int nDist = dist[ny][nx];
    			if(curDist > nDist) {
    				curDist = nDist;
    				ret = new Pair(ny,nx);
    			}
    		}
    		
    		return ret;
    	}
    	
    	/*
    	 * 메두사의 시선 
    	 */
    	public Sight findBestSight() {
    		Sight ret = NO_SIGHT;
    		
    		/*
    		 * sDir : 시선 방향 
    		 */
    		for(int sDir = 0; sDir < 4; sDir++) { 
    			Sight sight = new Sight(y,x,sDir);
    			if(sight.isHigher(ret))
    				ret = sight;
    		}
    		
    		return ret;
    	}
    }
    
    static int dy2[] = {-1,-1,0,1,1,1,0,-1};
    static int dx2[] = {0,1,1,1,0,-1,-1,-1};
    static class Tuple{
    	int y,x,dir;
    	public Tuple(int y,int x,int dir) {
    		this.y = y;
    		this.x = x;
    		this.dir = dir;
    	}
    }
    static Sight NO_SIGHT = new Sight();
    static class Sight{
    	int cnt, sDir;
    	int sighted[][];
    	List<Warrior> notFreezed;
    	
    	public Sight() {
    		this.cnt = -1;
    		this.sDir = 5;
    	}
    	/*
    	 * (N * N) / 4
    	 */
    	public Sight(int my, int mx, int sDir) {
    		this.sDir = sDir;
    		sighted = new int[N][N];
    		Queue<Tuple> q = new ArrayDeque<>();
    		int dir = sDir;
			if(sDir == 1)
    			dir = 4;
    		else if(sDir == 2)
    			dir = 6;
    		else if(sDir == 3)
    			dir = 2;
			
			List<Tuple> freezeds = new ArrayList<>();
			for(int nDir : getDirs(dir)) {
				int ny = my + dy2[nDir];
				int nx = mx + dx2[nDir];
				if(OOB(ny,nx))
					continue;
				sighted[ny][nx] = 1;
				q.add(new Tuple(ny,nx,nDir));
				if(wBoard[ny][nx] != 0) {
					freezeds.add(new Tuple(ny,nx, nDir));
				}
			}
			
			while(!q.isEmpty()) {
				Tuple cur = q.poll();
				
				for(int nDir : getDirs(cur.dir, sDir)) {
					int ny = cur.y + dy2[nDir];
					int nx = cur.x + dx2[nDir];
					
					if(OOB(ny,nx) || sighted[ny][nx] == 1)
						continue;
					if(wBoard[ny][nx] != 0)
						freezeds.add(new Tuple(ny,nx, nDir));
					
					q.add(new Tuple(ny,nx,nDir));
					sighted[ny][nx] = 1;
				}
			}
			
			for(Tuple freezed : freezeds) {
				if(sighted[freezed.y][freezed.x] == 2)
					continue;
				shadow(freezed);
			}
			
			
			notFreezed = new ArrayList<>();
			for(int i = 0; i < M; i++) {
				if(warriors[i] == NO_WARRIOR)
					continue;
				int wy = warriors[i].y;
				int wx = warriors[i].x;
				if(sighted[wy][wx] == 1) {
					cnt++;
					continue;
				}
				notFreezed.add(warriors[i]);
			}
			
			
    	}
    	
    	private void shadow(Tuple freezed) {
    		Queue<Tuple> q = new ArrayDeque<>();
    		for(int nDir : getDirs(freezed.dir, this.sDir)) {
    			int ny = freezed.y + dy2[nDir];
    			int nx = freezed.x + dx2[nDir];
    			if(OOB(ny,nx) || sighted[ny][nx] == 2)
    				continue;
    			q.add(new Tuple(ny,nx,nDir));
    			sighted[ny][nx] = 2;
    		}
    		
    		while(!q.isEmpty()) {
    			Tuple cur = q.poll();
    			
    			for(int nDir : getDirs(cur.dir, this.sDir)) {
    				int ny = cur.y + dy2[nDir];
    				int nx = cur.x + dx2[nDir];
    				
    				if(OOB(ny,nx) || sighted[ny][nx] == 2)
    					continue;
    				q.add(new Tuple(ny,nx,nDir));
    				sighted[ny][nx] = 2;
    			}
    		}
    		
    	}
    	
    	public boolean isHigher(Sight s) {
    		if(cnt != s.cnt)
    			return cnt > s.cnt;
    		return sDir < s.sDir;
    	}
    	
    	
    	private int[] getDirs(int dir) {
    		if(dir %2 == 0) {
    			return new int[] {(dir+7) % 8, dir, (dir + 1) % 8};
    		}
    		return new int[] {dir};
    	}
    	
    	private int[] getDirs(int cDir, int sDir) {
    		if(sDir == 0) {
    			if(cDir == 7) {
    				return new int[] {7, 0};
    			}else if(cDir == 0) {
    				return new int[] {0};
    			}else {
    				return new int[] {0, 1};
    			}
    			
    		}else if(sDir == 1) {
    			if(cDir == 3) {
    				return new int[] {3,4};
    			}else if(cDir == 4) {
    				return new int[] {4};
    			}else {
    				return new int[] {4,5};
    			}
    			
    		}else if(sDir == 2) {
    			if(cDir == 5) {
    				return new int[] {5,6};
    			}else if(cDir == 6) {
    				return new int[] {6};
    			}else {
    				return new int[] {6,7};
    			}
    		}else {
    			if(cDir == 1) {
    				return new int[] {1,2};
    			}else if(cDir == 2) {
    				return new int[] {2};
    			}else {
    				return new int[] {2,3};
    			}
    		}
    	}
    	

		
		public String toString() {
			return String.format("cnt: %d, sDir: %d", cnt, sDir);
		}
    }
    
    static Warrior NO_WARRIOR = new Warrior(-1,-1,-1);
    static class Warrior{
    	int id, y,x;
    	public Warrior(int id, int y,int x) {
    		this.id = id;
    		this.y = y;
    		this.x = x;
    	}
    	
    	public String toString() {
    		return String.format("id: %d | (%d, %d)", id, y,x);
    	}
    	
    	public Move move(int sighted[][]) {
    		int distance = getDistance(y,x, medusa.y, medusa.x);
    		int direction = -1;

    		
    		for(int dir = 0; dir < 4; dir++) {
    			int ny = y + dy[dir];
    			int nx = x + dx[dir];
    			
    			if(OOB(ny,nx) || sighted[ny][nx] == 1)
    				continue;
    			int D = getDistance(ny,nx, medusa.y, medusa.x);
    			if(distance > D) {
    				distance =  D;
    				direction = dir;
    			}
    		}
    		
    		if(direction == -1) {
    			return new Move(0, false);
    		}
    		
    		wBoard[y][x]--;
    		y = y + dy[direction];
    		x = x + dx[direction];
    		
    		
    		if(medusa.isSame(y,x)) {
    			warriors[id] = NO_WARRIOR;
    			return new Move(1,true);
    		}
    		wBoard[y][x]++;
    		
    		direction = -1;
    		for(int dir = 2; dir < 6; dir++) {
    			int nDir = dir % 4;
    			int ny = y + dy[nDir];
    			int nx = x + dx[nDir];
    			
    			if(OOB(ny,nx) || sighted[ny][nx] == 1)
    				continue;
    			int D = getDistance(ny,nx, medusa.y, medusa.x);
    			if(distance > D) {
    				distance = D;
    				direction = nDir;
    			}
    		}
    		
    		if(direction == -1)
    			return new Move(1, false);
    		
    		wBoard[y][x]--;
    		y = y + dy[direction];
    		x = x + dx[direction];
    		if(medusa.isSame(y,x)) {
    			warriors[id] = NO_WARRIOR;
    			return new Move(2, true);
    		}
    		wBoard[y][x]++;
    		return new Move(2,false);
    	}
    	
    }
    
    static class Move{
    	int dist;
    	boolean attack;
    	public Move(int dist, boolean attack) {
    		this.dist = dist;
    		this.attack = attack;
    	}
    }
    
    static int getDistance(int y1,int x1, int y2, int x2) {
    	return Math.abs(y1-y2) + Math.abs(x1-x2);
    }
    
    static int board[][];
    static int wBoard[][];
    
    static int dy[] = {-1,1,0,0};
    static int dx[] = {0,0,-1,1};
    
    static boolean OOB(int y,int x) {
    	return y<0 || y>=N || x < 0 || x>=N;
    }
    
    static void printBoard(int board[][]) {
    	for(int y=0; y<N; y++) {
    		for(int x= 0; x<N; x++) {
    			System.out.printf("%3d", board[y][x] == INF ? -1 : board[y][x]);
    		}
    		System.out.println();
    	}
    }
    
    /*
     * 메두사의 이동을 위한 초기 
     * 도착점을 기준으로 BFS
     * 그러면 각 칸에 공원으로의 최단거리가 새겨짐 
     */
	static void init() {
    	Queue<Pair> q = new ArrayDeque<>();
    	q.add(new Pair(ey,ex));
    	dist[ey][ex] = 0;
    	
    	while(!q.isEmpty()) {
    		Pair cur = q.poll();
    		
    		if(medusa.isSame(cur.y, cur.x))
    			break;
    		
    		for(int dir = 0; dir < 4; dir++) {
    			int ny = cur.y + dy[dir];
    			int nx = cur.x + dx[dir];
    			
    			if(OOB(ny,nx) || dist[ny][nx] != INF || board[ny][nx] == 1)
    				continue;
    			q.add(new Pair(ny,nx));
    			dist[ny][nx] = dist[cur.y][cur.x] + 1;
    		}
    	}
	}
	static int INF = Integer.MAX_VALUE / 2;
}

File: 250325/메두사와 전사들/medusa-and-warriors.py
import sys
from collections import deque

# 상수 및 전역 변수
INF = 10**9  # 충분히 큰 값
N = 0
M = 0
sy = sx = ey = ex = 0
board = []
wBoard = []
dist = []
medusa = None
warriors = []
NO_WARRIOR = None  # Warrior(-1, -1, -1) 로 초기화 후 할당

dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]

dy2 = [-1, -1, 0, 1, 1, 1, 0, -1]
dx2 = [0, 1, 1, 1, 0, -1, -1, -1]

class Pair:
    def __init__(self, y, x):
        self.y = y
        self.x = x
    def __str__(self):
        return f"({self.y}, {self.x})"

NO_PAIR = Pair(-1, -1)

class TupleX:
    # Java의 Tuple과 동일한 역할: y, x, 그리고 방향(dir)
    def __init__(self, y, x, d):
        self.y = y
        self.x = x
        self.dir = d

class Medusa:
    def __init__(self, y, x):
        self.y = y
        self.x = x

    def is_same(self, y, x):
        return self.y == y and self.x == x

    def end(self):
        global ey, ex
        return self.is_same(ey, ex)

    def move(self):
        nxt = self.find_nxt_pos()
        self.y = nxt.y
        self.x = nxt.x
        global warriors, wBoard, M, NO_WARRIOR
        for i in range(M):
            if warriors[i] == NO_WARRIOR:
                continue
            wy = warriors[i].y
            wx = warriors[i].x
            # 메두사와 전사의 좌표가 같으면 전사는 사라짐
            if self.is_same(wy, wx):
                wBoard[wy][wx] -= 1
                warriors[i] = NO_WARRIOR

    def find_nxt_pos(self):
        global dist, board, N
        curDist = dist[self.y][self.x]
        ret = NO_PAIR
        # 상하좌우 순서대로 다음 위치 후보를 찾음
        for dir in range(4):
            ny = self.y + dy[dir]
            nx = self.x + dx[dir]
            if OOB(ny, nx) or board[ny][nx] == 1:
                continue
            nDist = dist[ny][nx]
            if curDist > nDist:
                curDist = nDist
                ret = Pair(ny, nx)
        return ret

    def find_best_sight(self):
        best = NO_SIGHT
        for sDir in range(4):
            sight = Sight(self.y, self.x, sDir)
            if sight.is_higher(best):
                best = sight
        return best

class Sight:
    def __init__(self, my=None, mx=None, sDir=None):
        global N, wBoard, warriors, medusa
        # 기본 생성자: cnt=-1, sDir=5 (NO_SIGHT 용)
        if my is None or mx is None or sDir is None:
            self.cnt = -1
            self.sDir = 5
            self.sighted = None
            self.notFreezed = []
            return

        self.sDir = sDir
        # sighted 배열 초기화 (0으로 채움)
        self.sighted = [[0] * N for _ in range(N)]
        q = deque()
        # 시선 방향 조정: sDir 값에 따라 기준 방향 설정
        dir_val = sDir
        if sDir == 1:
            dir_val = 4
        elif sDir == 2:
            dir_val = 6
        elif sDir == 3:
            dir_val = 2

        freezeds = []
        for nDir in self.get_dirs_single(dir_val):
            ny = my + dy2[nDir]
            nx = mx + dx2[nDir]
            if OOB(ny, nx):
                continue
            self.sighted[ny][nx] = 1
            q.append(TupleX(ny, nx, nDir))
            if wBoard[ny][nx] != 0:
                freezeds.append(TupleX(ny, nx, nDir))
        while q:
            cur = q.popleft()
            for nDir in self.get_dirs(cur.dir, sDir):
                ny = cur.y + dy2[nDir]
                nx = cur.x + dx2[nDir]
                if OOB(ny, nx) or self.sighted[ny][nx] == 1:
                    continue
                if wBoard[ny][nx] != 0:
                    freezeds.append(TupleX(ny, nx, nDir))
                q.append(TupleX(ny, nx, nDir))
                self.sighted[ny][nx] = 1
        for freezed in freezeds:
            if self.sighted[freezed.y][freezed.x] == 2:
                continue
            self.shadow(freezed.y, freezed.x, dir_val)
        self.cnt = 0
        self.notFreezed = []
        for w in warriors:
            if w == NO_WARRIOR:
                continue
            wy = w.y
            wx = w.x
            if self.sighted[wy][wx] == 1:
                self.cnt += 1
            else:
                self.notFreezed.append(w)

    def shadow(self, y, x, dir_val):
        global N, medusa
        # medusa의 y 또는 x와 같으면 한 방향으로 그림자 투사
        if y == medusa.y or x == medusa.x:
            for dist_val in range(1, N + 1):
                ny = y + dy2[dir_val] * dist_val
                nx = x + dx2[dir_val] * dist_val
                if OOB(ny, nx):
                    continue
                self.sighted[ny][nx] = 2
        else:
            q = deque()
            q.append(Pair(y, x))
            self.sighted[y][x] = 2
            while q:
                cur = q.popleft()
                for nDir in self.get_dirs_xy(cur.y, cur.x, dir_val):
                    ny = cur.y + dy2[nDir]
                    nx = cur.x + dx2[nDir]
                    if OOB(ny, nx) or self.sighted[ny][nx] == 2:
                        continue
                    if self.sighted[ny][nx] == 1:
                        q.append(Pair(ny, nx))
                        self.sighted[ny][nx] = 2
            self.sighted[y][x] = 1

    def is_higher(self, s):
        # 우선 cnt가 높은 쪽, 같으면 sDir 값이 작은 쪽이 우선
        if self.cnt != s.cnt:
            return self.cnt > s.cnt
        return self.sDir < s.sDir

    def get_dirs_single(self, dir_val):
        if dir_val % 2 == 0:
            return [ (dir_val + 7) % 8, dir_val, (dir_val + 1) % 8 ]
        return [dir_val]

    def get_dirs_xy(self, wy, wx, dir_val):
        global medusa
        if dir_val == 0:
            if wy < medusa.y and wx < medusa.x:
                return [7, 0]
            if wy < medusa.y and wx > medusa.x:
                return [0, 1]
        elif dir_val == 2:
            if wy < medusa.y and wx > medusa.x:
                return [1, 2]
            if wy > medusa.y and wx > medusa.x:
                return [2, 3]
        elif dir_val == 4:
            if wy > medusa.y and wx > medusa.x:
                return [3, 4]
            if wy > medusa.y and wx < medusa.x:
                return [4, 5]
        elif dir_val == 6:
            if wy > medusa.y and wx < medusa.x:
                return [5, 6]
            if wy < medusa.y and wx < medusa.x:
                return [6, 7]
        return [dir_val]

    def get_dirs(self, cDir, sDir):
        if sDir == 0:
            if cDir == 7:
                return [7, 0]
            elif cDir == 0:
                return [0]
            else:
                return [0, 1]
        elif sDir == 1:
            if cDir == 3:
                return [3, 4]
            elif cDir == 4:
                return [4]
            else:
                return [4, 5]
        elif sDir == 2:
            if cDir == 5:
                return [5, 6]
            elif cDir == 6:
                return [6]
            else:
                return [6, 7]
        else:  # sDir == 3
            if cDir == 1:
                return [1, 2]
            elif cDir == 2:
                return [2]
            else:
                return [2, 3]

    def __str__(self):
        return f"cnt: {self.cnt}, sDir: {self.sDir}"

class Warrior:
    def __init__(self, id, y, x):
        self.id = id
        self.y = y
        self.x = x

    def __str__(self):
        return f"id: {self.id} | ({self.y}, {self.x})"

    def move(self, sighted):
        global medusa, wBoard, warriors, NO_WARRIOR
        distance = get_distance(self.y, self.x, medusa.y, medusa.x)
        direction = -1
        # 우선 상하좌우 중 이동
        for dir in range(4):
            ny = self.y + dy[dir]
            nx = self.x + dx[dir]
            if OOB(ny, nx) or sighted[ny][nx] == 1:
                continue
            D = get_distance(ny, nx, medusa.y, medusa.x)
            if distance > D:
                distance = D
                direction = dir
        if direction == -1:
            return Move(0, False)
        # 현재 위치에서 warrior 이동 및 wBoard 업데이트
        wBoard[self.y][self.x] -= 1
        self.y += dy[direction]
        self.x += dx[direction]
        if medusa.is_same(self.y, self.x):
            warriors[self.id] = NO_WARRIOR
            return Move(1, True)
        wBoard[self.y][self.x] += 1

        direction = -1
        # 추가 이동: 범위 2~5 (나머지 방향)
        for dir in range(2, 6):
            nDir = dir % 4
            ny = self.y + dy[nDir]
            nx = self.x + dx[nDir]
            if OOB(ny, nx) or sighted[ny][nx] == 1:
                continue
            D = get_distance(ny, nx, medusa.y, medusa.x)
            if distance > D:
                distance = D
                direction = nDir
        if direction == -1:
            return Move(1, False)
        wBoard[self.y][self.x] -= 1
        self.y += dy[direction]
        self.x += dx[direction]
        if medusa.is_same(self.y, self.x):
            warriors[self.id] = NO_WARRIOR
            return Move(2, True)
        wBoard[self.y][self.x] += 1
        return Move(2, False)

class Move:
    def __init__(self, dist, attack):
        self.dist = dist
        self.attack = attack

def get_distance(y1, x1, y2, x2):
    return abs(y1 - y2) + abs(x1 - x2)

def OOB(y, x):
    global N
    return y < 0 or y >= N or x < 0 or x >= N

def print_board(bd):
    global N, INF
    for y in range(N):
        for x in range(N):
            val = -1 if bd[y][x] == INF else bd[y][x]
            print(f"{val:3d}", end="")
        print()

def init():
    global ey, ex, dist, board, medusa, N
    q = deque()
    q.append(Pair(ey, ex))
    dist[ey][ex] = 0
    while q:
        cur = q.popleft()
        if medusa.is_same(cur.y, cur.x):
            break
        for dir in range(4):
            ny = cur.y + dy[dir]
            nx = cur.x + dx[dir]
            if OOB(ny, nx) or dist[ny][nx] != INF or board[ny][nx] == 1:
                continue
            q.append(Pair(ny, nx))
            dist[ny][nx] = dist[cur.y][cur.x] + 1

def move_all(best_sight):
    distSum = 0
    attackerCnt = 0
    for w in best_sight.notFreezed:
        move_result = w.move(best_sight.sighted)
        if move_result.attack:
            attackerCnt += 1
        distSum += move_result.dist
    return distSum, attackerCnt

def main():
    global N, M, sy, sx, ey, ex, board, wBoard, dist, medusa, warriors, INF, NO_WARRIOR
    input_data = sys.stdin.read().strip().split()
    if not input_data:
        return
    it = iter(input_data)
    N = int(next(it))
    M = int(next(it))
    sy = int(next(it))
    sx = int(next(it))
    ey = int(next(it))
    ex = int(next(it))
    # 메두사 초기화
    medusa_obj = Medusa(sy, sx)
    medusa = medusa_obj
    warriors = [None] * M
    wBoard = [[0] * N for _ in range(N)]
    for i in range(M):
        y = int(next(it))
        x = int(next(it))
        warriors[i] = Warrior(i, y, x)
        wBoard[y][x] += 1
    board = []
    dist = [[INF] * N for _ in range(N)]
    for y in range(N):
        row = [int(next(it)) for _ in range(N)]
        board.append(row)
    init()
    if dist[sy][sx] == INF:
        print(-1)
        return

    output_lines = []
    while True:
        medusa.move()
        if medusa.end():
            output_lines.append("0")
            break
        best_sight = medusa.find_best_sight()
        dsum, attacker_count = move_all(best_sight)
        output_lines.append(f"{dsum} {best_sight.cnt} {attacker_count}")
    print("\n".join(output_lines))

# NO_SIGHT 객체 (default Sight)
NO_SIGHT = Sight()
# NO_WARRIOR 전역 변수 초기화
NO_WARRIOR = Warrior(-1, -1, -1)

if __name__ == '__main__':
    main()
